go-controller/pkg/ovn/egressgw.go

```diff
@@ -197,99 +320,101 @@ func (oc *Controller) deletePodGWRoutesForNamespace(pod, namespace string) {
-				// clean up if there are no more routes for this podIP
-				if entry := nsInfo.podExternalRoutes[podIP]; len(entry) == 0 {
-					delete(nsInfo.podExternalRoutes, podIP)
-					// TODO (trozet): use the go bindings here and batch commands
-					// delete the ovn_cluster_router policy if the pod has no more exgws to revert back to normal
-					// default gw behavior
-					if err := oc.delHybridRoutePolicyForPod(net.ParseIP(podIP), node, false); err != nil {
-						klog.Error(err)
+					_, stderr, err := util.RunOVNNbctl("--if-exists", "--policy=src-ip",
+						"lr-route-del", gr, podIP+mask, gwIP.String())
+					if err != nil {
+						klog.Errorf("Unable to delete pod %s route to GR %s, GW: %s, stderr:%q, err:%v",
+							pod, gr, gwIP.String(), stderr, err)
+					} else {
+						klog.V(5).Infof("ECMP route deleted for pod: %s, on gr: %s, to gw: %s", pod,
+							gr, gwIP.String())
+
+						delete(routeInfo.routes[podIP], gwIP.String())
+						// clean up if there are no more routes for this podIP
+						if entry := routeInfo.routes[podIP]; len(entry) == 0 {
+							// TODO (trozet): use the go bindings here and batch commands
+							// delete the ovn_cluster_router policy if the pod has no more exgws to revert back to normal
+							// default gw behavior
+							if err := oc.delHybridRoutePolicyForPod(net.ParseIP(podIP), node, false); err != nil {
+								klog.Error(err)
+							}
+						}
 					}
+					cleanUpBFDEntry(gwIP.String(), gr, portPrefix)
 				}
 			}
-			cleanUpBFDEntry(gwIP.String(), gr, portPrefix)
+			routeInfo.Unlock()
 		}
 	}
-	delete(nsInfo.routingExternalPodGWs, pod)
 }
 
 // deleteGwRoutesForNamespace handles deleting all routes to gateways for a pod on a specific GR
 // This should only be called with a lock on nsInfo
-func (oc *Controller) deleteGWRoutesForNamespace(nsInfo *namespaceInfo) {
-	if nsInfo == nil {
-		return
-	}
-
+func (oc *Controller) deleteGWRoutesForNamespace(namespace string) {
 	// TODO(trozet): batch all of these with ebay bindings
-	for podIP, gwToGr := range nsInfo.podExternalRoutes {
-		for gw, gr := range gwToGr {
-			if utilnet.IsIPv6String(gw) != utilnet.IsIPv6String(podIP) {
-				continue
-			}
-			mask := GetIPFullMask(podIP)
-			node := util.GetWorkerFromGatewayRouter(gr)
-			if err := oc.delHybridRoutePolicyForPod(net.ParseIP(podIP), node, false); err != nil {
-				klog.Error(err)
-			}
-			_, stderr, err := util.RunOVNNbctl("--if-exists", "--policy=src-ip",
-				"lr-route-del", gr, podIP+mask, gw)
-			if err != nil {
-				klog.Errorf("Unable to delete src-ip route to GR router, stderr:%q, err:%v", stderr, err)
-			} else {
-				delete(nsInfo.podExternalRoutes, podIP)
-			}
+	routeInfos := oc.getRouteInfosForNamespace(namespace)
+	for _, routeInfo := range routeInfos {
+		for podIP, gwToGr := range routeInfo.routes {
+			for gw, gr := range gwToGr {
+				if utilnet.IsIPv6String(gw) != utilnet.IsIPv6String(podIP) {
+					continue
+				}
```

Comment:
deletePodGWRoutesForNamespace() checks for gr != "", do we need to do that here? If not, do we need to do it in the other function?

Reply:
good point. I dont think we should ever be in a situation where there is an entry in the cache with an empty gr. Let's say we should check for that condition just to be safe...then in that case if it is empty, we should remove that value from the cache because it is invalid. I'm going to work on a separate PR to address the differences in the logic between these two functions, and maybe try to consolidate it to a single function.